tests/adapters/test_protocol.py

```diff
@@ -22,10 +22,177 @@
 # THE SOFTWARE.
 #
 
-from pymeasure.adapters.protocol import ProtocolAdapter
+from pymeasure.adapters.protocol import to_bytes, ProtocolAdapter
+
+from pytest import mark, raises, fixture
+
+
+@mark.parametrize("input, output", (("superXY", b"superXY"),
+                                    ([1, 2, 3, 4], b"\x01\x02\x03\x04"),
+                                    (5, b"5"),
+                                    (4.6, b"4.6")
+                                    ))
+def test_to_bytes(input, output):
+    assert to_bytes(input) == output
+
+
+def test_to_bytes_invalid():
+    with raises(TypeError):
+        to_bytes(5.5j)
 
 
 def test_protocol_instantiation():
-    a = ProtocolAdapter()
-    assert not a
+    a = ProtocolAdapter([("write", "read"), ("write_only",)])
+    assert a.comm_pairs == [("write", "read"), ("write_only",)]
     # TODO: Make this first test assert something useful
+
+
+class Test_write:
+    @mark.parametrize("command, written", (("hey Ho", b"hey Ho"),
+                                           (b"superX", "superX")
+                                           ))
+    def test_write(self, command, written):
+        a = ProtocolAdapter([(written, 5)])
+        a.write(command)
+        assert a._read_buffer == b"5"
+
+    def test_write_without_response(self):
+        a = ProtocolAdapter([("Hey ho",)])
+        a.write("Hey ho")
+        assert a._read_buffer == b""
+
+    def test_wrong_command(self):
+        a = ProtocolAdapter([("Hey ho",)])
+        with raises(AssertionError):
+            a.write("Something different")
+
+    def test_not_enough_pairs(self):
+        a = ProtocolAdapter([("a",)])
+        a._index = 1
+        with raises(IndexError):
+            a.write("b")
+
+
+class Test_write_bytes:
+    @fixture(scope="class")
+    def written1(self):
+        """Write in a single turn."""
+        a = ProtocolAdapter([("written", 5)])
+        a.write_bytes(b"written")
+        return a
+
+    @fixture(scope="class")
+    def written2(self):
+        """Write in two turns."""
+        a = ProtocolAdapter([("written", 5)])
+        a.write_bytes(b"writ")
+        a.write_bytes(b"ten")
+        return a
+
+    def test_write_write_buffer1(self, written1):
+        assert written1._write_buffer == b""
+
+    def test_write_write_buffer2(self, written2):
+        assert written2._write_buffer == b""
```

Comment:
To avoid the duplicated tests, you could parametrize the fixture itself: https://docs.pytest.org/en/6.2.x/fixture.html#parametrizing-fixtures

Reply:
I tried parametrizing the tests, which does not work with fixtures. Thanks for showing me the parametrization of fixtures.